File: lib/languages/kotlin/memory-management.kt
// Memory management in Kotlin is handled by the JVM's garbage collector

// Basic memory allocation
fun basicAllocation() {
    // Memory is automatically allocated
    val list = ArrayList<Int>(1000000) // Allocates memory for a list with initial capacity
    
    // Initialize the list
    for (i in 0 until 1000000) {
        list.add(i)
    }
    
    // Memory will be automatically freed by the garbage collector
    // when it's no longer referenced
}

// Stack vs Heap allocation
fun stackVsHeap() {
    // Primitive types are often optimized by the JVM
    val stackInt = 42 // May be allocated on the stack
    
    // Objects are allocated on the heap
    val heapObject = Object() // Allocated on the heap
    
    // The JVM uses escape analysis to optimize allocations,
    // but generally all objects go on the heap initially
}

// Memory leaks can still happen
class MemoryLeakExample {
    // Static references can cause memory leaks
    companion object {
        // This list will never be garbage collected as long as the class is loaded
        private val leakyList = mutableListOf<String>()
        
        fun addToList(item: String) {
            leakyList.add(item)
        }
    }
    
    // Inner classes hold implicit references to their outer class
    inner class Leaky {
        fun doSomething() {
            // This inner class holds a reference to the outer class,
            // preventing it from being garbage collected
        }
    }
    
    // Solution: Use static nested classes instead of inner classes
    class NotLeaky {
        fun doSomething() {
            // This class doesn't hold a reference to the outer class
        }
    }
}

// Weak references to prevent memory leaks
fun weakReferences() {
    // Strong reference - prevents garbage collection
    var strongRef = Object()
    
    // Weak reference - doesn't prevent garbage collection
    val weakRef = WeakReference(strongRef)
    
    // When strongRef is set to null, the object can be garbage collected
    // even though weakRef still points to it
    strongRef = null
    
    // weakRef.get() may return null if the object has been collected
    println("Weak reference is null: ${weakRef.get() == null}")
}

// Soft references for memory-sensitive caching
fun softReferences() {
    // Soft references are cleared at the discretion of the garbage collector,
    // typically in response to memory pressure
    val cache = mutableMapOf<String, SoftReference<ByteArray>>()
    
    // Add a large object to the cache
    val data = ByteArray(1024 * 1024) // 1MB
    cache["key"] = SoftReference(data)
    
    // The garbage collector may clear this reference if memory is needed
    val cachedData = cache["key"]?.get()
}

// Value classes to reduce memory overhead
@JvmInline
value class UserId(val value: Int)

fun valueClasses() {
    // Without value classes, this would create a new object on the heap
    val regularId = UserId(123)
    
    // With @JvmInline, the UserId is compiled to just the underlying Int
    // in most cases, avoiding object allocation
    processUser(regularId)
}

fun processUser(id: UserId) {
    println("Processing user: ${id.value}")
}

// Object pools to reduce GC pressure
object StringPool {
    private val pool = mutableMapOf<String, String>()
    
    fun intern(s: String): String {
        return pool.getOrPut(s) { s }
    }
}

fun objectPooling() {
    // Without pooling, these create two different objects
    val str1 = "Hello, World!"
    val str2 = "Hello, " + "World!"
    
    // With pooling, we can reuse the same object
    val pooled1 = StringPool.intern(str1)
    val pooled2 = StringPool.intern(str2)
    
    println("Same reference: ${pooled1 === pooled2}")
}

// Using arrays for performance-critical code
fun arrays() {
    // ArrayList is more convenient but has more overhead
    val list = ArrayList<Int>(1000)
    
    // IntArray is more efficient for primitive types
    val array = IntArray(1000)
    
    // Fill both
    for (i in 0 until 1000) {
        list.add(i)
        array[i] = i
    }
    
    // IntArray avoids boxing/unboxing overhead
}

// String concatenation efficiency
fun stringEfficiency() {
    // Inefficient: creates multiple intermediate String objects
    var result = ""
    for (i in 1..100) {
        result += i.toString()
    }
    
    // More efficient: StringBuilder
    val builder = StringBuilder()
    for (i in 1..100) {
        builder.append(i)
    }
    val efficientResult = builder.toString()
    
    println("Length: ${efficientResult.length}")
}

// Sequence operations to avoid intermediate collections
fun sequences() {
    val numbers = (1..1000000).toList()
    
    // This creates multiple intermediate collections
    val result1 = numbers
        .filter { it % 2 == 0 }
        .map { it * 2 }
        .take(10)
    
    // This uses sequences to avoid intermediate collections
    val result2 = numbers.asSequence()
        .filter { it % 2 == 0 }
        .map { it * 2 }
        .take(10)
        .toList()
    
    println("Results equal: ${result1 == result2}")
}

// Main function to demonstrate memory management
fun main() {
    basicAllocation()
    stackVsHeap()
    weakReferences()
    softReferences()
    valueClasses()
    objectPooling()
    arrays()
    stringEfficiency()
    sequences()
}

File: lib/languages/go/memory-management.go
// Memory management in Go is handled by a garbage collector

// Basic memory allocation
func basicAllocation() {
	// Memory is automatically allocated
	slice := make([]int, 1000000) // Allocates memory for 1 million integers
	
	// Initialize the slice
	for i := range slice {
		slice[i] = i
	}
	
	// Memory will be automatically freed by the garbage collector
	// when it's no longer referenced
}

// Stack vs Heap allocation
func stackVsHeap() {
	// Stack allocation (fast, automatically freed when function returns)
	var stackArray [10]int
	stackArray[0] = 42
	
	// Heap allocation (managed by GC)
	heapSlice := make([]int, 10)
	heapSlice[0] = 42
	
	// Go's compiler uses escape analysis to determine whether to allocate
	// on the stack or heap, but generally:
	// - Large objects go on the heap
	// - Objects whose address is taken may go on the heap
	// - Objects that outlive the function go on the heap
}

// Memory leaks can still happen
func potentialLeak() {
	// This is a common pattern that can cause memory leaks
	ch := make(chan int)
	
	go func() {
		// If this goroutine is blocked forever, the channel and goroutine
		// will never be garbage collected
		ch <- 42
	}()
	
	// If we never read from the channel, the goroutine will be blocked
	// and both the goroutine and channel will leak
	// val := <-ch
}

// Controlling GC behavior
func controlGC() {
	// Force garbage collection
	runtime.GC()
	
	// Get current memory statistics
	var stats runtime.MemStats
	runtime.ReadMemStats(&stats)
	
	fmt.Printf("Allocated memory: %d bytes\n", stats.Alloc)
	fmt.Printf("Total memory allocated: %d bytes\n", stats.TotalAlloc)
	fmt.Printf("System memory: %d bytes\n", stats.Sys)
	fmt.Printf("GC cycles: %d\n", stats.NumGC)
}

// Memory pooling to reduce GC pressure
func memoryPooling() {
	// Create a pool of byte slices
	var bufferPool = sync.Pool{
		New: func() interface{} {
			buffer := make([]byte, 1024)
			return &buffer
		},
	}
	
	// Get a buffer from the pool
	buffer := bufferPool.Get().(*[]byte)
	
	// Use the buffer
	for i := 0; i < 100; i++ {
		(*buffer)[i] = byte(i)
	}
	
	// Return the buffer to the pool when done
	bufferPool.Put(buffer)
	
	// This reduces GC pressure by reusing objects
	// instead of allocating new ones
}

// Using value types to reduce allocations
func valueTypes() {
	// Value types (structs) are allocated on the stack when possible
	type Point struct {
		X, Y int
	}
	
	// This doesn't allocate on the heap
	p := Point{X: 10, Y: 20}
	
	// This allocates on the heap
	pHeap := &Point{X: 10, Y: 20}
	
	// Using value types for small, short-lived objects
	// can significantly reduce GC pressure
	_ = p
	_ = pHeap
}

// Setting GC percentage
func tuneGC() {
	// Set the GC target percentage
	// Higher values mean less frequent GC but more memory usage
	// Default is 100
	old := debug.SetGCPercent(200) // GC will run when memory doubles
	
	fmt.Printf("Old GC percentage: %d\n", old)
	
	// You can also set this with GOGC environment variable
}

// Preallocating memory to avoid reallocations
func preallocation() {
	// Bad: causes multiple reallocations as the slice grows
	var slice []int
	for i := 0; i < 10000; i++ {
		slice = append(slice, i)
	}
	
	// Good: preallocate with the expected capacity
	slice2 := make([]int, 0, 10000)
	for i := 0; i < 10000; i++ {
		slice2 = append(slice2, i)
	}
}

// Using strings efficiently
func stringEfficiency() {
	// Strings are immutable in Go
	s := "hello"
	
	// This creates a new string and allocates memory
	s += " world"
	
	// More efficient for multiple concatenations
	var builder strings.Builder
	builder.WriteString("hello")
	builder.WriteString(" ")
	builder.WriteString("world")
	result := builder.String()
	
	_ = result
}

// Main function to demonstrate memory management
func main() {
	basicAllocation()
	stackVsHeap()
	// Avoid the leak
	// potentialLeak()
	controlGC()
	memoryPooling()
	valueTypes()
	tuneGC()
	preallocation()
	stringEfficiency()
}

